### ISSUE TITLE
Construction of with expression in `lowering/byte_operators.cpp` does not match getters in `with_exprt` [rfc]

### ISSUE THREAD

**Issue by thomasspriggs** [diffblue/cbmc#7272]
This section constructs a with expression containing multiple where/new value pairs -
https://github.com/diffblue/cbmc/blob/8bd2617310877c27c912faedd2b33d5d8d8e2116/src/solvers/lowering/byte_operators.cpp#L1514-L1517
However the API here only provides access to a single where/new value pair - 
https://github.com/diffblue/cbmc/blob/8bd2617310877c27c912faedd2b33d5d8d8e2116/src/util/std_expr.h#L2444-L2462

What should be the preferred way to deal with this mismatch?
  A. Update the lowering to nest with expressions rather than constructing a single expression with multiple pairs.
  B. Update `with_exprt` to expose the possibility of multiple pairs.
  C. Work around the mismatch for the moment using `exprt::operands` to access the additional pairs.
I am not a huge fan of option C as it leaves the possibility of multiple pairs relatively surprising when referring to `with_exprt`. Option A seems relatively preferable as it keeps the API simpler, however I have also presented option B in case there are performance reasons for preferring more operands over more deeply nested expressions.

**Comment by tautschnig** [comment]
I believe "B" is the way to go: the reason "A" isn't the best approach is that it could result in deeply nested expressions when a large array is to be updated in this way.

**Comment by thomasspriggs** [comment]
> I believe "B" is the way to go: the reason "A" isn't the best approach is that it could result in deeply nested expressions when a large array is to be updated in this way.

For context, we came across this issue whilst working on SMT support. The closest array theory equivalent term seems to be the `store` function - https://smtlib.cs.uiowa.edu/theories-ArraysEx.shtml This array theory function only supports a single where/new value pair. So for the purposes of SMT I would expect this to be converted into a deeply nested expression which ever choice we make for the API here.

I appreciate there may be benefits to avoiding deeply nested expressions for SAT support however.

**Comment by tautschnig** [comment]
Yes, I suppose that nested expression is unavoidable at SMT level at this time.

**Comment by tautschnig** [comment]
Labelling this "Version 6:" I believe we should reconsider the use of `with_exprt` altogether, much like the comment at its declaration suggests, or else restructure the operands within the expression such that each operand has two child operands. Else it will be very difficult to assess whether all readers of such expressions correctly process all updates that a `with` expression seeks to denote.

Any such change will result in (GOTO-) binary incompatibility, and we should postpone that until the major-version bump.
